### PR TITLE
feat: add single tailnet support to moons

### PR DESCRIPTION
**Pull request by coadler**
This PR adds support for the `single_tailnet` experiment to moons. Instead of a websocket for each agent connection, it reuses the `MultiAgent` paradigm backed by a single websocket.

### CODE REVIEW

File: enterprise/coderd/coderd.go [hunk 1]
```diff
@@ -395,35 +405,39 @@ func (api *API) updateEntitlements(ctx context.Context) error {
 		return nil
 	}
 
-	featureChanged := func(featureName codersdk.FeatureName) (changed bool, enabled bool) {
+	featureChanged := func(featureName codersdk.FeatureName) (initial, changed, enabled bool) {
```
Comment (coadler) [inline enterprise/coderd/coderd.go:408]: Everything below was pulled into a separate pr https://github.com/coder/coder/pull/8586

File: enterprise/tailnet/workspaceproxy.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+)
+
+func ServeWorkspaceProxy(conn net.Conn, ma agpl.MultiAgentConn) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	go func() {
+		err := forwardNodesToWorkspaceProxy(ctx, conn, ma)
+		if err != nil {
+			cancel()
```
Comment (deansheather) [inline enterprise/tailnet/workspaceproxy.go:23]: This doesn't seem to do anything, should it close the conn?

File: enterprise/tailnet/workspaceproxy.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/enterprise/wsproxy/wsproxysdk"
+	agpl "github.com/coder/coder/tailnet"
+)
+
+func ServeWorkspaceProxy(conn net.Conn, ma agpl.MultiAgentConn) error {
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (deansheather) [inline enterprise/tailnet/workspaceproxy.go:17]: Should be based on request context

File: enterprise/wsproxy/wsproxysdk/wsproxysdk_test.go [hunk 4]
```diff
@@ -1,21 +1,37 @@
...
+	"github.com/google/uuid"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"golang.org/x/xerrors"
+	"nhooyr.io/websocket"
+	"tailscale.com/types/key"
 
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/slogtest"
+
```
Comment (deansheather) [inline enterprise/wsproxy/wsproxysdk/wsproxysdk_test.go:27]: should merge import groups

**Review by deansheather — APPROVED** [review]

**Comment by deansheather** [comment]
You commented `Everything below was pulled into a separate pr https://github.com/coder/coder/pull/8586` but it looks like it's all still there

File: enterprise/tailnet/workspaceproxy.go (hunk 2 above)
Comment (coadler) [reply enterprise/tailnet/workspaceproxy.go:23]: Good point

**Comment by coadler** [comment]
> You commented `Everything below was pulled into a separate pr https://github.com/coder/coder/pull/8586` but it looks like it's all still there

I pulled it out to get merged separately. My tests won't pass without it so I kept it in here until I could get it merged.
